components/suggest/src/suggestion.rs

```diff
@@ -3,19 +3,34 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
+use std::fmt::Write;
+
+use chrono::{Datelike, Local, Timelike};
+
+/// The template parameter for a timestamp in a "raw" sponsored suggestion URL.
+const TIMESTAMP_TEMPLATE: &str = "%YYYYMMDDHH%";
+
+/// The length, in bytes, of a timestamp in a "cooked" sponsored suggestion URL.
+///
+/// Cooked timestamps don't include the leading or trailing `%`, so this is
+/// 2 bytes shorter than [`TIMESTAMP_TEMPLATE`].
+const TIMESTAMP_LENGTH: usize = 10;
+
 /// A suggestion from the database to show in the address bar.
 #[derive(Clone, Debug, Eq, Hash, PartialEq)]
 pub enum Suggestion {
     Amp {
         title: String,
         url: String,
+        raw_url: String,
```

Comment:
Worth noting that there can be no template in the raw urls for certain advertisers, meaning that `url` and `raw_url` will be the same for those. Though the vast majority of suggestions do have that, so I think we can live with that tiny overhead for now.

Reply:
Definitely! I made these `Optional<String>` at first, and only passed them if `url` and `raw_url` are different—but I think always passing them makes things a bit easier for consumers, at the cost of some overhead for us. Otherwise, our consumers would need to do something like `suggestion.rawUrl ?? suggestion.url` to get at the raw URL.